File: tests/AcceptanceTests/StepDefinitions/AdminDashboardSteps.cs
// ============================================================================
// RAJ Financial - Admin Dashboard Step Definitions
// ============================================================================
// Reqnroll step definitions for AdminDashboard.feature
// ============================================================================

using Microsoft.Playwright;
using RajFinancial.AcceptanceTests.Hooks;
using Reqnroll;

namespace RajFinancial.AcceptanceTests.StepDefinitions;

/// <summary>
/// Step definitions for admin dashboard scenarios.
/// </summary>
[Binding]
public class AdminDashboardSteps
{
    private readonly ScenarioContext scenarioContext;
    private IPage Page => scenarioContext.GetPage();

    protected AdminDashboardSteps(ScenarioContext scenarioContext)
    {
        this.scenarioContext = scenarioContext;
    }

    [Then(@"I should see statistics cards with numeric values")]
    public async Task ThenIShouldSeeStatisticsCardsWithNumericValues()
    {
        // Wait for page to settle and cards to render
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
        await Page.WaitForTimeoutAsync(500);

        var cards = Page.Locator(".card h3");
        await Assertions.Expect(cards).ToHaveCountAsync(4, new() { Timeout = 15000 });

        var count = await cards.CountAsync();
        Assert.True(count >= 4, "Should have at least four statistic values");

        for (int i = 0; i < Math.Min(4, count); i++)
        {
            var text = (await cards.Nth(i).InnerTextAsync()).Trim();
            // Normalize currency symbols and commas
            var normalized = text.Replace(",", string.Empty)
                                  .Replace("$", string.Empty)
                                  .Replace("€", string.Empty)
                                  .Replace("£", string.Empty)
                                  .Trim();

            Assert.True(decimal.TryParse(normalized, out _), $"Card {i + 1} should display a numeric value but was '{text}'");
        }
    }

    [Then(@"all statistics should have labels")]
    public async Task ThenAllStatisticsShouldHaveLabels()
    {
        // Statistics should be in .card elements
        var cards = await Page.Locator(".card").AllAsync();
        Assert.True(cards.Count >= 4, "Should have at least 4 statistics cards");

        foreach (var card in cards.Take(4))
        {
            var text = await card.TextContentAsync();
            Assert.NotNull(text);
            Assert.True(text.Length > 5, "Each card should have descriptive text");
        }
    }

    [Then(@"all statistics should have icons")]
    public async Task ThenAllStatisticsShouldHaveIcons()
    {
        // Check for icons (using Open Iconic or other icon libraries)
        var icons = await Page.Locator(".oi, svg, i[class*='icon']").CountAsync();
        Assert.True(icons >= 4, "Should have at least 4 icons for statistics");
    }

    [Then(@"I should see activity items in the recent activity section")]
    public async Task ThenIShouldSeeActivityItemsInTheRecentActivitySection()
    {
        var activityItems = await Page.Locator("li").CountAsync();
        Assert.True(activityItems >= 1, "Recent activity should have items");
    }

    [Then(@"the statistics cards should be stacked vertically")]
    public async Task ThenTheStatisticsCardsShouldBeStackedVertically()
    {
        // On mobile, grid should be single column
        var isStacked = await Page.EvaluateAsync<bool>(@"
            () => {
                const grid = document.querySelector('.grid');
                if (!grid) return false;
                const styles = window.getComputedStyle(grid);
                // Check if grid-template-columns is set to 1 column
                return styles.gridTemplateColumns === '1fr' ||
                       styles.gridTemplateColumns.startsWith('repeat(1,');
            }
        ");

        // Alternative check: verify cards are stacked by checking their positions
        if (!isStacked)
        {
            var cards = await Page.Locator(".card").AllAsync();
            if (cards.Count >= 2)
            {
                var box1 = await cards[0].BoundingBoxAsync();
                var box2 = await cards[1].BoundingBoxAsync();

                // Cards should be vertically stacked (y positions different)
                Assert.NotNull(box1);
                Assert.NotNull(box2);
                Assert.True(box2.Y > box1.Y, "Cards should be stacked vertically on mobile");
            }
        }
    }

    [Then(@"the quick actions should be visible")]
    public async Task ThenTheQuickActionsShouldBeVisible()
    {
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
        await Page.WaitForTimeoutAsync(500);

        var quickActions = Page.Locator("[data-testid='quick-actions'], h5:has-text(\"Quick Actions\")");
        await quickActions.First.WaitForAsync(new() { Timeout = 20000, State = WaitForSelectorState.Attached });
        await quickActions.First.ScrollIntoViewIfNeededAsync();
        await Assertions.Expect(quickActions.First).ToBeVisibleAsync(new() { Timeout = 20000 });
    }

    [Then(@"the statistics cards should adapt to tablet layout")]
    public async Task ThenTheStatisticsCardsShouldAdaptToTabletLayout()
    {
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
        await Page.WaitForTimeoutAsync(500);

        var cards = Page.Locator(".grid .card");
        await Assertions.Expect(cards).ToBeVisibleAsync();
    }

    [Then(@"all statistics should have accessible labels")]
    public async Task ThenAllStatisticsShouldHaveAccessibleLabels()
    {
        // Check that statistics have proper labels (h6, labels, or aria-labels)
        var cards = await Page.Locator(".card").AllAsync();

        foreach (var card in cards.Take(4))
        {
            var hasLabel = await card.Locator("h1, h2, h3, h4, h5, h6, label, [aria-label]").CountAsync();
            Assert.True(hasLabel >= 1, "Each statistic card should have an accessible label");
        }
    }

    [Then(@"all icons should be hidden from screen readers or have labels")]
    public async Task ThenAllIconsShouldBeHiddenFromScreenReadersOrHaveLabels()
    {
        var icons = await Page.Locator(".oi, svg, i[class*='icon']").AllAsync();

        foreach (var icon in icons)
        {
            var ariaHidden = await icon.GetAttributeAsync("aria-hidden");
            var ariaLabel = await icon.GetAttributeAsync("aria-label");
            var title = await icon.GetAttributeAsync("title");

            var isAccessible = ariaHidden == "true" ||
                              !string.IsNullOrWhiteSpace(ariaLabel) ||
                              !string.IsNullOrWhiteSpace(title);

            Assert.True(isAccessible, "Icons should be hidden from screen readers or have labels");
        }
    }
}

File: tests/e2e/step-definitions/admin-dashboard.steps.ts
// ============================================================================
// Admin Dashboard Step Definitions
// ============================================================================
// Steps for AdminDashboard.feature.
// NOTE: Admin routes (/admin/dashboard) are not yet implemented in React.
// These steps are scaffolded for when the admin pages are built.
// ============================================================================

import { Then } from "@cucumber/cucumber";
import { expect } from "@playwright/test";
import { CustomWorld } from "../support/world";

Then(
  "I should see the admin dashboard",
  async function (this: CustomWorld) {
    await this.page.waitForLoadState("networkidle");
    await this.page.waitForTimeout(1000);

    const dashboard = this.page.locator("[data-testid='admin-dashboard']");
    const title = this.page.locator(
      "[data-testid='admin-dashboard-title'], h1:has-text('Administrator Dashboard')"
    );

    let found = false;
    try {
      await expect(dashboard).toBeVisible({ timeout: 15000 });
      found = true;
    } catch {
      try {
        await expect(title).toBeVisible({ timeout: 5000 });
        found = true;
      } catch {
        // neither found
      }
    }
    expect(found).toBe(true);
  }
);

Then(
  "I should not see an access denied message",
  async function (this: CustomWorld) {
    const content = await this.page.content();
    expect(content).not.toContain("Access Denied");
    expect(content.toLowerCase()).not.toContain("not authorized");
  }
);

Then(
  "I should see an access denied message",
  async function (this: CustomWorld) {
    const content = await this.page.content();
    expect(
      content.includes("Access Denied") ||
        content.toLowerCase().includes("not authorized")
    ).toBe(true);
  }
);

Then(
  "I should see activity items in the recent activity section",
  async function (this: CustomWorld) {
    const items = await this.page.locator("li").count();
    expect(items).toBeGreaterThanOrEqual(1);
  }
);

Then(
  "the statistics cards should be stacked vertically",
  async function (this: CustomWorld) {
    const cards = await this.page.locator(".card").all();
    if (cards.length >= 2) {
      const box1 = await cards[0].boundingBox();
      const box2 = await cards[1].boundingBox();
      expect(box1).not.toBeNull();
      expect(box2).not.toBeNull();
      expect(box2!.y).toBeGreaterThan(box1!.y);
    }
  }
);

Then(
  "the quick actions should be visible",
  async function (this: CustomWorld) {
    await this.page.waitForLoadState("networkidle");
    const quickActions = this.page.locator(
      "[data-testid='quick-actions'], h5:has-text('Quick Actions')"
    );
    await quickActions.first().scrollIntoViewIfNeeded();
    await expect(quickActions.first()).toBeVisible({ timeout: 20000 });
  }
);

Then(
  "the statistics cards should adapt to tablet layout",
  async function (this: CustomWorld) {
    await this.page.waitForLoadState("networkidle");
    const cards = this.page.locator(".grid .card");
    await expect(cards.first()).toBeVisible();
    expect(await cards.count()).toBeGreaterThanOrEqual(1);
  }
);

Then(
  "all statistics should have accessible labels",
  async function (this: CustomWorld) {
    const cards = await this.page.locator(".card").all();
    for (const card of cards.slice(0, 4)) {
      const labels = await card
        .locator("h1, h2, h3, h4, h5, h6, label, [aria-label]")
        .count();
      expect(labels).toBeGreaterThanOrEqual(1);
    }
  }
);

Then(
  "all icons should be hidden from screen readers or have labels",
  async function (this: CustomWorld) {
    const icons = await this.page.locator("svg, i[class*='icon']").all();
    for (const icon of icons) {
      const ariaHidden = await icon.getAttribute("aria-hidden");
      const ariaLabel = await icon.getAttribute("aria-label");
      const title = await icon.getAttribute("title");

      expect(
        ariaHidden === "true" ||
          (ariaLabel?.trim().length ?? 0) > 0 ||
          (title?.trim().length ?? 0) > 0
      ).toBe(true);
    }
  }
);
